### ISSUE TITLE
Infinite loop in layout with recursive dependency

### ISSUE THREAD

**Issue by tronical** [sixtyfpsui/sixtyfps#117]
The following example grows indefinitely:

```
export Test := Window {
    GridLayout {
        padding: 10px;
        Row {
            Rectangle {
                height: root.height;
            }
        }
    }
}
```

It would seem that the layout's height is calculated by taking the height of the window, adding the default padding/spacing and setting a new width and height on the window. Once the window manager applied the size, the layout will be dirty and re-calculated again.

Edit: Decision on how to fix: https://github.com/sixtyfpsui/sixtyfps/issues/117#issuecomment-777308667

**Comment by matklad** [comment]
Hit this one as well!

**Comment by ogoffart** [comment]
The  example in the issue is now working, most likely same as https://github.com/sixtyfpsui/sixtyfps/issues/140

Could you re-open a new issue with a new test case if it can still be reproduced?  Thanks!

**Comment by matklad** [comment]
https://user-images.githubusercontent.com/1711539/107091615-220ee180-6813-11eb-850c-8de2af7edde6.mp4

https://github.com/matklad/120wpm/commit/5f0aa2e48797c84e4630399821c893101b667ef2

**Comment by ogoffart** [comment]
The behavior depends on the backend actually, The Qt backend seems to not re-compute more often, So it requires change to re-compute the layout, while the GL backend seem to recompute the layout at each event loop iteration.
This testcase reproduce the problem also with Qt when pressing keys.

```css
MainWindow := Window {
    HorizontalLayout {
        padding: 10px;
        Rectangle {
            // BUG: this tells the layout to have a minimum-size = 10px*2 + size: so there is a constraint loop
            width: 100%;
            height: 100%;
        }
    }

    // The rest is just code so that it recomputes the layout when pressing a key
    property <int> foo;
    forward-focus: input;
    input := FocusScope {
        key-pressed(key) => { foo += 1; accept }
        Rectangle {
            // just some dummy things in a layout to force re-computation when it changes
            VerticalLayout {  Rectangle { width: 1px * foo; }   }
        }
    }
}
```

I'm unsure how to "fix" this. Should the compiler tries to detect such loop and error or warns when this happen? or what would the expected behavior from this?

**Comment by ogoffart** [comment]
Note that detecting loop reliably might require a full property dependency analysis, as the property might depends on other properties:

```css
MainWindow := Window {
    HorizontalLayout {
        padding: 10px;
        Rectangle {
            property <length> indirection: root.width;
            width: indirection;
        }
    }
   //...
```

**Comment by ogoffart** [comment]
One way to resolve this: If the height or width property reference the parent property, they should not restrict the minimum or maximum size (as if: they could stretch)

**Comment by ogoffart** [comment]
Decision: If the property contains a reference to the parent size, this will be an error detected once https://github.com/sixtyfpsui/sixtyfps/issues/169 is done.   If the property contains a literal percentage (eg: `width: 50%`) then this does not constraint the maximum and minimum size
